install-tools.sh

```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+# This script is created based on Dockerfile.
+
+set -x
+
+apt-get -q update && apt-get upgrade -y
+apt-get install -y --no-install-recommends -t bullseye-backports \
+      git make ccache cpio curl gnupg universal-ctags cscope socat patch \
+      gperf quilt bmake byacc python3-pip device-tree-compiler tmux \
+      libmpfr6 libfdt1 libpython3.9 libsdl2-2.0-0 libglib2.0-0 libpixman-1-0
+apt-key adv --fetch-keys https://apt.llvm.org/llvm-snapshot.gpg.key
+echo "deb http://apt.llvm.org/bullseye/ llvm-toolchain-bullseye-14 main" > \
+      /etc/apt/sources.list.d/llvm-14.list && apt-get update
+apt-get install -y --no-install-recommends -t bullseye-backports \
+      clang-14 clang-format-14 llvm-14 lld-14
+
+# patch & quilt required by lua and programs in contrib/
+# gperf required by libterminfo
+# socat & tmux required by launch
+
+pip3 install setuptools wheel && pip3 install -r requirements.txt
+curl -O https://mimiker.ii.uni.wroc.pl/download/mipsel-mimiker-elf_latest_amd64.deb && \
+    dpkg -i mipsel-mimiker-elf_latest_amd64.deb && rm -f mipsel-mimiker-elf_latest_amd64.deb
+curl -O https://mimiker.ii.uni.wroc.pl/download/aarch64-mimiker-elf_latest_amd64.deb && \
+    dpkg -i aarch64-mimiker-elf_latest_amd64.deb && rm -f aarch64-mimiker-elf_latest_amd64.deb
+curl -O https://mimiker.ii.uni.wroc.pl/download/riscv32-mimiker-elf_latest_amd64.deb && \
+    dpkg -i riscv32-mimiker-elf_latest_amd64.deb && rm -f riscv32-mimiker-elf_latest_amd64.deb
+curl -O https://mimiker.ii.uni.wroc.pl/download/riscv64-mimiker-elf_latest_amd64.deb && \
+    dpkg -i riscv64-mimiker-elf_latest_amd64.deb && rm -f riscv64-mimiker-elf_latest_amd64.deb
+curl -O https://mimiker.ii.uni.wroc.pl/download/qemu-mimiker_latest_amd64.deb && \
+    dpkg -i qemu-mimiker_latest_amd64.deb && rm -f qemu-mimiker_latest_amd64.deb
```

Comment:
This is **completely** wrong!

- you assume that host is running with bullseye-backport Debian **without** checking that in **any** way, you can do that in well known container that we are building in dockerfiles
- auto _yes_ can cause a lot of damage for host system
- where `bmake` is used by our build system?
- libs should be autodeps for other packages
- so what is a source of llvm packages? bullseye-backports or external repo?
- who should run this script?
  - user? in this case you don't have permission to modify /usr and /etc so you can't install anything in that way
  - root? downloading files as root is a brilliant idea; in this case you also run pip globally which is easy way to seriously break your system - you will end-up with python libs in random state overwriting those provided by package manager

Reply:
Could you suggest some changes? I have no idea how to fix things that you have mentioned.

Reply:
something like this? first you check privileges, next some preconditions, upgrade to root for packages management, calculate distro codename (it's todo), again packages management, local pip3 installation and at the end install custom toolchain if it's supported by distro

```
#!/bin/bash

function fatal() {
    echo $1
    exit 1
}

if [ "`id -u`" -eq 0 ]; then
    echo "you are running as root... it is not supported"
    exit 1
fi

[ -f /etc/apt/sources.list.d/llvm.list ] && fatal 'you already have llvm sources.list'
[ -f requirements.txt ] || fatal 'you are not in mimiker repo topdir'

LLVM_VERSION=14
DEBS="git make ccache cpio curl gnupg universal-ctags cscope socat patch gperf quilt byacc python3 python3-pip python3-virtualenv device-tree-compiler tmux lsb-release gdb-multiarch qemu-system-aarch64"

DEBS_LLVM="clang-$LLVM_VERSION clang-format-$LLVM_VERSION llvm-$LLVM_VERSION lld-$LLVM_VERSION"

[ -f /usr/bin/apt ] || fatal '/usr/bin/apt is missing; is your distro debian based?'
[ -f /usr/bin/sudo ] || fatal 'sudo is missing; (su -c "apt install sudo")'

echo "install required packages"
echo "upgrade privileges to root user"
echo "install packages from base repository"
sudo -u root sh -c "apt-get update -q && apt install $DEBS"
echo "prepare sources.list for llvm && install llvm-$LLVM_VERSION"

DISTRO=$(lsb_release -is 2>/dev/null)
VERSION=$(lsb_release -sr 2>/dev/null)
# fetch codename
source /etc/os-release

# calculate codename
CODENAME="bullseye"

REPO="deb http://apt.llvm.org/${CODENAME}/ llvm-toolchain-bullseye-${LLVM_VERSION} main"
sudo -u root sh -c "apt-key adv --fetch-keys https://apt.llvm.org/llvm-snapshot.gpg.key && echo $REPO > /etc/apt/sources.list.d/llvm.list && apt-get update -q && apt-get install $DEBS_LLVM"

echo "drop privileges to $USER"

echo "install local python dependencies"
python3 -m venv mimiker-env
. mimiker-env/bin/activate
pip3 install -r requirements.txt

case $CODENAME in
    bullseye)
    WORKDIR=$(mktemp -d)
        echo "$WORKDIR created"

        echo "download prebuild toolchain"
        curl -o $WORKDIR/mipsel-mimiker-elf_latest_amd64.deb https://mimiker.ii.uni.wroc.pl/download/mipsel-mimiker-elf_latest_amd64.deb
        curl -o $WORKDIR/aarch64-mimiker-elf_latest_amd64.deb https://mimiker.ii.uni.wroc.pl/download/aarch64-mimiker-elf_latest_amd64.deb
        curl -o $WORKDIR/riscv32-mimiker-elf_latest_amd64.deb https://mimiker.ii.uni.wroc.pl/download/riscv32-mimiker-elf_latest_amd64.deb
        curl -o $WORKDIR/riscv64-mimiker-elf_latest_amd64.deb https://mimiker.ii.uni.wroc.pl/download/riscv64-mimiker-elf_latest_amd64.deb
        curl -o $WORKDIR/qemu-mimiker_latest_amd64.deb https://mimiker.ii.uni.wroc.pl/download/qemu-mimiker_latest_amd64.deb

        echo "install toolchain"
        sudo -u root sh -c "dpkg -i $WORKDIR/*.deb"

        echo "cleanup $WORKDIR"
        rm -r $WORKDIR
        ;;
    *)
        echo "your distribution doesn't support our prebuild toolchain"
        ;;
esac

exit 0
```

Reply:
Updated. There are only distributions left to add (e.g. 'sid' and some ubuntu distributions)